### PR TITLE
Patchscopes implementation and Initial Tests

### PR DESCRIPTION
**Pull request by FergusFettes**
Making a PR of a first pass at patchscopes to have somewhere to discuss it. I didn't get token identity working nicely so I would like to add more high level unit tests.

### CODE REVIEW

**Review by tvhong — COMMENTED** [review]
This is a good first draft of the PatchScope class.

I don't think we have the right interface for this class in this PR as the class has too many responsibilities. But, we can always refactor after implementing a few more lenses.

File: obvspython/patchscopes.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    REMOTE: bool = False
+
+    _source_hidden_state: torch.Tensor = field(init=False)
+    _mapped_hidden_state: torch.Tensor = field(init=False)
+    _target_outputs: List[torch.Tensor] = field(init=False, default_factory=list)
+
+    def __post_init__(self):
+        # Load models
+        self.source_model = LanguageModel(self.source.model_name, device_map=self.source.device)
+        if self.source.model_name == self.target.model_name:
```
[the remark this answers is not in the document]
Comment (FergusFettes) [reply obvspython/patchscopes.py:112]: hmm good idea

File: obvspython/patchscopes.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    def __repr__(self):
+        return (
+            f"TargetContext(prompt={self.prompt}, position={self.position}, "
+            f"model_name={self.model_name}, layer={self.layer}, device={self.device}, "
+            f"max_new_tokens={self.max_new_tokens}, "
+            f"mapping_function={self.mapping_function})"
+        )
+
+
+@dataclass
```
[the remark this answers is not in the document]
Comment (FergusFettes) [reply obvspython/patchscopes.py:94]: hmm, im ambivalent about this. i like having the documentation of all the expected class information at the top

File: obvspython/patchscopes.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        self.source_model = LanguageModel(self.source.model_name, device_map=self.source.device)
+        if self.source.model_name == self.target.model_name:
+            self.target_model = self.source_model
+        else:
+            self.target_model = LanguageModel(self.target.model_name, device_map=self.target.device)
+
+        self.tokenizer = self.source_model.tokenizer
+        self.get_position()
+
+    def source_forward_pass(self, source: Optional[SourceContext] = None):
```
[the remark this answers is not in the document]
Comment (FergusFettes) [reply obvspython/patchscopes.py:120]: yeah fair enough

File: tests/test_patchscopes.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
+from obvspython.patchscopes import SourceContext, TargetContext, Patchscope
+
+import torch
+import numpy as np
+
+
+class TestContext:
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (FergusFettes) [reply tests/test_patchscopes.py:8]: this is just a style question in the case of TestClasses. If you remove staticmethod you have to add self to them all. Doesnt matter either way

File: obvspython/patchscopes.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
+# Implementation of the patchscopes framework: https://arxiv.org/abs/2401.06102
+# Patchscopes takes a representation like so:
+# - (S, i, M, ℓ) corresponds to the source from which the original hidden representation is drawn.
+#   - S is the source input sequence.
+#   - i is the position within that sequence.
```
[the remark this answers is not in the document]
Comment (FergusFettes) [reply obvspython/patchscopes.py:5]: Thats true. I'm unsure how to deal with this. Its useful to have a range of positions in some cases, but its kind of an extension of the framework. Maybe ill just make a comment here.

File: obvspython/patchscopes_base.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
+from typing import Optional
+from copy import deepcopy
+from abc import ABC, abstractmethod
+
+import torch
+
+
+class PatchscopesBase(ABC):
```
[the remark this answers is not in the document]
Comment (FergusFettes) [reply obvspython/patchscopes_base.py:8]: makes sense for the abstract methods. though in this case they are all none lol

File: obvspython/patchscopes_base.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+    @property
+    def _n_layers_gpt(self):
+        return len(self.target_model.transformer.h)
+
+    @property
+    def _n_layers_llama2(self):
+        return len(self.target_model.model.layers)
+
+    def get_activation_pair(self, string_a: str, string_b: Optional[str] = None):
```
[the remark this answers is not in the document]
Comment (FergusFettes) [reply obvspython/patchscopes_base.py:162]: yes fair enough. ill remove these for now. they are useful for activation steering but we can add them back later

File: tests/test_patchscopes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tvhong) [reply tests/test_patchscopes.py:8]: Interesting. I'll look into this a bit more once I get to write some tests.

**Pull request merged**